File: android/lib/src/main/java/duck/cameras/android/service/CameraFinder.java
package duck.cameras.android.service;

import android.content.res.Resources;
import android.util.Log;

import java.net.Inet4Address;
import java.net.Inet6Address;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

import duck.cameras.android.R;
import duck.cameras.android.model.Callback;
import duck.cameras.android.model.Camera;
import duck.cameras.android.model.Result;
import duck.cameras.android.model.XmlNode;
import duck.cameras.android.util.DeviceUtils;

public class CameraFinder {
    private static final int LOCAL_PORT = 55000;
    private static final int UPNP_PORT = 3702;
    private static final int UPNP_TIMEOUT = 1000;
    private static final InetSocketAddress UPNP_IPv4 = new InetSocketAddress("239.255.255.250", UPNP_PORT);
    private static final InetSocketAddress UPNP_IPv6 = new InetSocketAddress("[FF05::C]", UPNP_PORT);

    private final ResourceLoader resourceLoader;

    public CameraFinder(Resources resources) {
        this.resourceLoader = new ResourceLoader(resources);
    }

    public void findAsync(Callback<List<Camera>> callback) {
        new Thread(() -> {
            try {
                final ArrayList<Camera> result = new ArrayList<>();

                if (DeviceUtils.isRunningOnEmulator()) {
                    String response = resourceLoader.loadString(R.raw.ws_emulator_response);
                    result.addAll(getCameras(Collections.singletonList(response)));
                } else {
                    final String uuid = UUID.randomUUID().toString();
                    final String data = resourceLoader.loadString(R.raw.ws_discovery_probe, uuid);

                    for (InetAddress localAddress : NetworkService.getSiteLocalAddresses()) {
                        if (localAddress instanceof Inet6Address) {
                            continue;
                        }
                        InetSocketAddress localEndPoint = new InetSocketAddress(localAddress, LOCAL_PORT);
                        InetSocketAddress remoteEndPoint = localAddress instanceof Inet4Address ? UPNP_IPv4 : UPNP_IPv6;
                        List<String> responses = NetworkService.udpResquest(localEndPoint, remoteEndPoint, data, UPNP_TIMEOUT);
                        result.addAll(getCameras(responses));
                    }
                }

                callback.execute(Result.ok(result));
            } catch (RuntimeException e) {
                Log.e(CameraFinder.class.getSimpleName(), "findAsync error", e);
                callback.execute(Result.error(Collections.emptyList(), e));
            }
        }).start();
    }

    private List<Camera> getCameras(List<String> responses) {
        final ArrayList<Camera> result = new ArrayList<>();
        for (String response : responses) {
            XmlNode envelope = XmlParser.parse(response);
            XmlNode probeMatch = envelope.get("Body").get("ProbeMatches").get("ProbeMatch");
            if (probeMatch.get("Types").value().equals("dn:NetworkVideoTransmitter")) {
                String xAddrs = probeMatch.get("XAddrs").value();
                Camera camera = new Camera();
                camera.endPoint = NetworkService.getEndpoint(xAddrs);
                getServices(camera);
                getInformation(camera);
                getProfiles(camera);
                result.add(camera);
            }
        }
        return result;
    }

    private void getServices(Camera camera) {
        String url = "http://" + camera.endPoint + "/onvif/device_service";
        String response = NetworkService.httpPost(url, resourceLoader.loadString(R.raw.ws_get_services));
        XmlNode envelope = XmlParser.parse(response);
        Iterable<XmlNode> services = envelope.get("Body").get("GetServicesResponse").getAll("Service");
        for (XmlNode serviceNode : services) {
            Camera.Service service = new Camera.Service();
            service.namespace = serviceNode.get("Namespace").value();
            service.address = serviceNode.get("XAddr").value();
            service.version = serviceNode.get("Version").get("Major").value()
                    + "." + serviceNode.get("Version").get("Minor").value();
            camera.services.add(service);
        }
    }

    private void getInformation(Camera camera) {
        String url = "http://" + camera.endPoint + "/onvif/device_service";
        String response = NetworkService.httpPost(url, resourceLoader.loadString(R.raw.ws_get_device_information));
        XmlNode envelope = XmlParser.parse(response);
        XmlNode deviceInfo = envelope.get("Body").get("GetDeviceInformationResponse");
        camera.information.manufacturer = deviceInfo.get("Manufacturer").value();
        camera.information.model = deviceInfo.get("Model").value();
        camera.information.firmwareVersion = deviceInfo.get("FirmwareVersion").value();
        camera.information.serialNumber = deviceInfo.get("SerialNumber").value();
        camera.information.hardwareId = deviceInfo.get("HardwareId").value();
    }

    private void getProfiles(Camera camera) {
        String url = "http://" + camera.endPoint + "/onvif/device_service";
        String response = NetworkService.httpPost(url, resourceLoader.loadString(R.raw.ws_get_profiles));
        XmlNode envelope = XmlParser.parse(response);
        Iterable<XmlNode> profiles = envelope.get("Body").get("GetProfilesResponse").getAll("Profiles");
        for (XmlNode profileNode : profiles) {
            Camera.Profile profile = new Camera.Profile();
            profile.name = profileNode.get("Name").value();
            profile.token = profileNode.attr("token");
            profile.streamUri = getStreamUri(camera, profile.token);
            profile.snapshotUri = getSnapshotUri(camera, profile.token);
            camera.profiles.add(profile);
        }
    }

    private String getStreamUri(Camera camera, String profileToken) {
        String url = "http://" + camera.endPoint + "/onvif/Media";
        String response = NetworkService.httpPost(url, resourceLoader.loadString(R.raw.ws_get_stream_uri, profileToken));
        XmlNode envelope = XmlParser.parse(response);
        return envelope.get("Body").get("GetStreamUriResponse").get("MediaUri").get("Uri").value();
    }

    private String getSnapshotUri(Camera camera, String profileToken) {
        String url = "http://" + camera.endPoint + "/onvif/Media";
        String response = NetworkService.httpPost(url, resourceLoader.loadString(R.raw.ws_get_snapshot_uri, profileToken));
        XmlNode envelope = XmlParser.parse(response);
        return envelope.get("Body").get("GetSnapshotUriResponse").get("MediaUri").get("Uri").value();
    }
}

File: windows/Service/CameraFinder.cs
using Duck.Cameras.Windows.Model;
using Duck.Cameras.Windows.Properties;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Net.Sockets;
using System.Threading.Tasks;
using System.Xml.Linq;

namespace Duck.Cameras.Windows.Service
{
    public class CameraFinder
    {
        private const int localPort = 55000;
        private const int upnpPort = 3702;
        private const int upnpTimeout = 1000;
        private readonly IPEndPoint upnpIPv4 = new IPEndPoint(IPAddress.Parse("239.255.255.250"), upnpPort);
        private readonly IPEndPoint upnpIPv6 = new IPEndPoint(IPAddress.Parse("[FF05::C]"), upnpPort);

        public Task<IEnumerable<Camera>> FindAsync()
        {
            return Task.Run<IEnumerable<Camera>>(async () =>
            {
                List<Camera> result = new List<Camera>();

                string uuid = Guid.NewGuid().ToString();
                string data = MessageLoader.Load(Resources.ws_discovery_probe, uuid);

                foreach (var localAddress in NetworkService.GetSiteLocalAddresses())
                {
                    if (localAddress.AddressFamily == AddressFamily.InterNetwork)
                    {
                        var localEndPoint = new IPEndPoint(localAddress, localPort);
                        var remoteEndPoint = upnpIPv4;
                        var responses = NetworkService.UdpRequest(localEndPoint, remoteEndPoint, data, upnpTimeout);
                        result.AddRange(await GetCameras(responses));
                    }
                }

                return result;
            });
        }

        private async Task<IEnumerable<Camera>> GetCameras(IEnumerable<string> responses)
        {
            List<Camera> result = new List<Camera>();

            foreach (var response in responses)
            {
                var probeMatch = SoapParser.Parse(response).Get("ProbeMatches").Get("ProbeMatch");
                if (probeMatch.Get("Types").Value == "dn:NetworkVideoTransmitter")
                {
                    string xAddrs = probeMatch.Get("XAddrs").Value;
                    Camera camera = new Camera();
                    camera.EndPoint = new Uri(xAddrs).Authority;
                    await GetServices(camera);
                    await GetInformation(camera);
                    await GetProfiles(camera);
                    result.Add(camera);
                }
            }

            return result;
        }

        private async Task GetServices(Camera camera)
        {
            var url = "http://" + camera.EndPoint + "/onvif/device_service";
            var response = await NetworkService.HttpPostAsync(url, MessageLoader.Load(Resources.ws_get_services));

            var services = SoapParser.Parse(response).Get("GetServicesResponse").GetAll("Service");
            foreach (XElement serviceNode in services)
            {
                CameraService service = new CameraService();
                service.Namespace = serviceNode.Get("Namespace").Value;
                service.Address = serviceNode.Get("XAddr").Value;
                service.Version = serviceNode.Get("Version").Get("Major").Value
                        + "." + serviceNode.Get("Version").Get("Minor").Value;
                camera.Services.Add(service);
            }
        }

        private async Task GetInformation(Camera camera)
        {
            var url = "http://" + camera.EndPoint + "/onvif/device_service";
            var response = await NetworkService.HttpPostAsync(url, MessageLoader.Load(Resources.ws_get_device_information));

            var deviceInfo = SoapParser.Parse(response).Get("GetDeviceInformationResponse");
            camera.Information.Manufacturer = deviceInfo.Get("Manufacturer").Value;
            camera.Information.Model = deviceInfo.Get("Model").Value;
            camera.Information.FirmwareVersion = deviceInfo.Get("FirmwareVersion").Value;
            camera.Information.SerialNumber = deviceInfo.Get("SerialNumber").Value;
            camera.Information.HardwareId = deviceInfo.Get("HardwareId").Value;
        }

        private async Task GetProfiles(Camera camera)
        {
            var url = "http://" + camera.EndPoint + "/onvif/device_service";
            var response = await NetworkService.HttpPostAsync(url, MessageLoader.Load(Resources.ws_get_profiles));

            var profiles = SoapParser.Parse(response).Get("GetProfilesResponse").GetAll("Profiles");
            foreach (var profileNode in profiles)
            {
                CameraProfile profile = new CameraProfile();
                profile.Name = profileNode.Get("Name").Value;
                profile.Token = profileNode.Attribute("token").Value;
                profile.StreamUri = await GetStreamUri(camera, profile.Token);
                profile.SnapshotUri = await GetSnapshotUri(camera, profile.Token);
                camera.Profiles.Add(profile);
            }
        }

        private async Task<string> GetStreamUri(Camera camera, string profileToken)
        {
            var url = "http://" + camera.EndPoint + "/onvif/Media";
            var response = await NetworkService.HttpPostAsync(url, MessageLoader.Load(Resources.ws_get_stream_uri, profileToken));

            return SoapParser.Parse(response).Get("GetStreamUriResponse").Get("MediaUri").Get("Uri").Value;
        }

        private async Task<string> GetSnapshotUri(Camera camera, string profileToken)
        {
            var url = "http://" + camera.EndPoint + "/onvif/Media";
            var response = await NetworkService.HttpPostAsync(url, MessageLoader.Load(Resources.ws_get_snapshot_uri, profileToken));

            return SoapParser.Parse(response).Get("GetSnapshotUriResponse").Get("MediaUri").Get("Uri").Value;
        }

        public async Task<IEnumerable<Camera>> FindFromSettingsAsync(bool update)
        {
            LoginToken loginToken = LocalSettingsManager.LoadLoginToken();
            RemoteSettings settings = await RemoteSettingsLoader.LoadAsync(update);
            await NetworkService.SetMode(settings);

            List<Camera> result = new List<Camera>();

            foreach (var endPoint in settings.EndPoints)
            {
                try
                {
                    Camera camera = new Camera();
                    camera.EndPoint = endPoint.Host + ":" + endPoint.LocalPort.Command;
                    camera.Name = endPoint.Name;
                    CameraProfile profile = new CameraProfile();
                    profile.Token = endPoint.Profile;
                    profile.StreamUri = CreateStreamUri(endPoint, loginToken);
                    profile.SnapshotUri = CreateSnapshotUri(endPoint, loginToken);
                    camera.Profiles.Add(profile);
                    result.Add(camera);
                }
                catch (Exception e)
                {                    
                    Debug.Write(e, nameof(CameraFinder.FindFromSettingsAsync));
                }
            }

            return result;
        }

        private string CreateStreamUri(RemoteSettingsEndPoint endPoint, LoginToken loginToken)
        {
            string path = endPoint.StreamPath.Replace("{token}", loginToken.Value);
            return string.Format("rtsp://{0}:{1}/{2}", endPoint.Host, endPoint.LocalPort.Stream, path);
        }

        private string CreateSnapshotUri(RemoteSettingsEndPoint endPoint, LoginToken loginToken)
        {
            string path = endPoint.SnapshotPath.Replace("{token}", loginToken.Value);
            return string.Format("http://{0}:{1}/{2}", endPoint.Host, endPoint.LocalPort.Snapshot, path);
        }
    }
}
